Docs: Add docs for the Dagster facade (#3736)
TobikoData/sqlmesh: docs/cloud/features/scheduler/dagster.md

- Importing the `SQLMeshEnterpriseDagster` class
- Creating a `SQLMeshEnterpriseDagster` instance configured with the environment variables you declared above
- Creating the `Definitions` object with the `create_definitions()` method

In the `definitions.py` file in your project, insert the following:

```python title="definitions.py" linenums="1"
from dagster import EnvVar
from tobikodata.scheduler_facades.dagster import SQLMeshEnterpriseDagster

sqlmesh = SQLMeshEnterpriseDagster(
    url=EnvVar("TOBIKO_CLOUD_BASE_URL").get_value(),
    token=EnvVar("TOBIKO_CLOUD_TOKEN").get_value(),
)

tobiko_cloud_definitions = sqlmesh.create_definitions(environment="prod")
```

!!! info
    If there is an existing definitions object already declared in your Dagster project, you can merge in the Tobiko Cloud definitions like so:

    ```python
    defs = Definitions(...)

    defs.merge(tobiko_cloud_definitions)
    ```

This is all that's needed to integrate with Tobiko Cloud!

Once Dagster loads your project, you will see some new objects become available.

## Available Dagster objects

The Dagster integration exports the following:

- An `Asset` object for every SQLMesh Model

>   ![Dagster UI Asset Lineage](./dagster/asset_lineage.png)

Reviewer: copy and paste this image up top after the intro sentence. Get the reader excited about the outcome they want to achieve.

- On the first run, it will trigger the `sync` job. This synchronises the materialization status of all Models and Audits from Tobiko Cloud to their corresponding Assets / Asset Checks in Dagster
- On subsequent runs, if there is a new Cloud Scheduler run, it will trigger the `mirror` job to mirror the outcome of that run locally.

![Job run records in the Dagster UI](./dagster/job_run_records.png)

!!! info "Why are there two jobs?"
    The Tobiko Cloud scheduler is optimised to prevent doing unnecessary work. In any given cadence run, it's possible only a subset of models are updated.
    Therefore, there would be no materialization information for the remaining models. There is also no materialization information for objects that are never part of a cadence run, such as [seeds](../../../concepts/models/seed_models.md).

    The `sync` jobs exists to copy across the current state of every object in the project as a base for the `mirror` job to build on.

If you need to manually refresh materialization information for all models, you can run the `sync` job manually from the Dagster UI:
![Manually run the sync job](./dagster/manual_sync_run.png)

> ## How it works

Reviewer: Move this section towards the top before the setup section.

Reviewer: yes

> Tobiko Cloud uses a custom approach to Dagster integration - this section describes how it works.

Reviewer: ```suggestion
Tobiko Cloud uses a custom approach to Dagster integration.
```

Author: I just copied this from the [Airflow](https://sqlmesh.readthedocs.io/en/stable/cloud/features/scheduler/airflow/#how-it-works) page, should it be changed there too?

iirc @treysp added it

> The Mirror job mirrors the progress of the Tobiko Cloud scheduler run. Each local task reflects the outcome of its corresponding remote task. If an asset is materialized remotely, the job emits a Dagster materialization event.

Reviewer: ```suggestion
The `Mirror` job mirrors the progress of the Tobiko Cloud scheduler run. Each local task reflects the outcome of its corresponding remote task. If an asset is materialized remotely, the job emits a Dagster materialization event.
```

Author: I made this change as `mirror` instead of `Mirror` to keep consistency with the other usages in the document

> This allows you to observe at a glance how your data pipeline is progressing, displayed alongside your other pipelines in Dagster. No need to navigate to Tobiko Cloud!

Reviewer: ```suggestion
This allows you to observe at a glance how your data pipeline is progressing, displayed alongside your other pipelines in Dagster. No need to context switch to Tobiko Cloud!

All scheduling configurations are set per SQLMesh model as a [`cron`](https://sqlmesh.readthedocs.io/en/stable/concepts/models/overview/?h=cron#cron), so Dagster's mechanism for [declarative scheduling](https://dagster.io/blog/declarative-scheduling) is not in scope. 
```

> ### Why a custom approach?

Reviewer: move this similar to the how it works section up top

Tobiko Cloud's scheduler performs multiple optimizations to ensure that your pipelines run correctly and efficiently. Those optimizations are only possible within our SQLMesh-aware scheduler.

Our approach allows you to benefit from those optimizations while retaining the flexibility to attach extra tasks or logic to the Assets created by Tobiko Cloud in your broader pipeline orchestration context.

> Because `run`s are still triggered by the Tobiko Cloud scheduler and tasks in the local DAG just reflect their remote equivalent in Tobiko Cloud, we call our custom approach a *facade*.

Reviewer: We can talk about this later, but I recommend we call it a `mirror` integration. Facade has negative connotations and isn't clear what it means in plain English.

Author: Does that mean we are changing the name of this concept and also need to update the Airflow page?

We would also need to change the package name on pypi

Once the Code Location is reloaded, the `mirror` job executes as usual with the updated definitions.

### Manual method

At any time, you can click the "Reload" button against the Code Location in Dagster. This will pull through any new Asset and Asset Check definitions from Tobiko Cloud.

![Dagster reload code location](./dagster/reload_code_location.png)

## Attaching custom logic

Dagster includes a robust events system. You can listen for events on any of the Assets or Jobs exposed by our Dagster integration and respond to them by running your own custom logic.

To listen for materialization events on Assets, you can use an [Asset Sensor](https://docs.dagster.io/concepts/partitions-schedules-sensors/asset-sensors).

To listen for job runs, you can use a [Run Status Sensor](https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#run-status-sensors).

Dagster also provides a framework called [Declarative Automation](https://docs.dagster.io/concepts/automation/declarative-automation) that builds on top of these.

### Examples

Here are some examples of running custom logic in response to Tobiko Cloud events.

> Note that Dagster has a lot of flexibility in how it can be configured and the methods we describe below arent necessarily the right choice for every configuration. We recommend familiarizing yourself with Dagster's [Automation](https://docs.dagster.io/concepts/automation) features to get the most out of your Tobiko Cloud deployment with Dagster.

Reviewer: ```suggestion
Note that Dagster has a lot of flexibility in how it can be configured and the methods we describe below aren't necessarily the right choice for every configuration. We recommend familiarizing yourself with Dagster's [Automation](https://docs.dagster.io/concepts/automation) features to get the most out of your Tobiko Cloud deployment with Dagster.
```

> #### Response to run status

Reviewer: add a picture of what it looks like in the Dagster UI if anything. I don't know off the cuff how they visually represent this in the UI.

Author: There isnt a UI visualization per se, unlike Airflow where you modify the DAG we create to hang off extra tasks.

It would just add a standalone entry to the list of jobs?

Reviewer: can you show that image of the standalone entry? the visual validation helps readers.

Author: Ok, added

Here's an example that triggers a Slack notification:

```python
from dagster import run_status_sensor, job, DagsterRunStatus, EnvVar, RunRequest, RunStatusSensorContext
from dagster_slack import SlackResource

mirror_job = tobiko_cloud_definitions.get_job_def("tobiko_cloud_mirror_run_prod")

@op
def slack_op(slack: SlackResource):
    #note: see the dagster-slack docs here: https://docs.dagster.io/_apidocs/libraries/dagster-slack
    slack.get_client().chat_postMessage(channel="#notifications", ...)

@job
def notify_slack():
    slack_op()

@run_status_sensor(
    run_status=DagsterRunStatus.STARTED,
    monitored_jobs=[mirror_job],
    request_job=notify_slack
)
def on_tobiko_cloud_start_run(context: RunStatusSensorContext):
    return RunRequest()
```

!!! info

    If your Tobiko Cloud definitions live in their own code location, you can use a [JobSelector](https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#cross-code-location-run-status-sensors) to target the Mirror job instead of the direct reference.

> #### Response to Asset Materialization

Reviewer: add a picture of what it should look and feel like in the UI

Author: Added

```python
@run_status_sensor(
    run_status=DagsterRunStatus.SUCCESS,
    monitored_jobs=[mirror_job],
    request_job=send_email_job
)
def on_tobiko_cloud_start_run(context: RunStatusSensorContext):
    return RunRequest()
```

You can adjust the `run_status` argument to listen for different statuses, the `monitored_jobs` argument to monitor other jobs and the `request_job` argument to trigger a different job.

Here's an example that triggers a Slack notification:

```python
from dagster import run_status_sensor, job, DagsterRunStatus, EnvVar, RunRequest, RunStatusSensorContext
from dagster_slack import SlackResource

mirror_job = tobiko_cloud_definitions.get_job_def("tobiko_cloud_mirror_run_prod")

@op
def slack_op(slack: SlackResource):
    #note: see the dagster-slack docs here: https://docs.dagster.io/_apidocs/libraries/dagster-slack
    slack.get_client().chat_postMessage(channel="#notifications", ...)

@job
def notify_slack():
    slack_op()

@run_status_sensor(
    run_status=DagsterRunStatus.STARTED,
    monitored_jobs=[mirror_job],
    request_job=notify_slack
)
def on_tobiko_cloud_start_run(context: RunStatusSensorContext):
    return RunRequest()
```

> ![Dagster run status sensor](./dagster/run_status_sensor.png)

Reviewer: move this image up top so it's below this #### Response to run status header

similar reasoning to the top-level image. Get someone excited to create that outcome

> ![Dagster run status sensor](./dagster/run_status_sensor.png)

Reviewer: move this under the subheader: Response to Asset Materialization

First, navigate to your Dagster project or [create a new one](https://docs.dagster.io/guides/build/projects/creating-a-new-project).

Next, add the `tobiko-cloud-scheduler-facade` library to the `install_requires` section of your [Dagster project](https://docs.dagster.io/guides/understanding-dagster-project-files)'s `setup.py`:

```python title="setup.py" hl_lines="3"
install_requires=[
    "dagster",
    "tobiko-cloud-scheduler-facade[dagster]"
],
```

### Connect Dagster to Tobiko Cloud

Dagster recommends [injecting secret values using Environment Variables](https://docs.dagster.io/guides/dagster/using-environment-variables-and-secrets#using-environment-variables-and-secrets). The exact method you should use depends on how your organization deploys Dagster.

On this page, we demonstrate the secrets method Dagster recommends for **local development**.

In your Dagster project, create an `.env` file if it does not already exist. Next, specify environment variables containing the Tobiko Cloud URL and token secrets:

```sh title=".env"
TOBIKO_CLOUD_BASE_URL=<URL for your Tobiko Cloud project> # ex: https://cloud.tobikodata.com/sqlmesh/tobiko/public-demo/

> TOBIKO_CLOUD_TOKEN=<your Tobiko Cloud API token> # ex: ioawjioefja1

Reviewer: ```suggestion
TOBIKO_CLOUD_TOKEN=<your Tobiko Cloud API token> # ex: 'ioawjioefja1'
```

single quotes are important for unique characters
